### ISSUE TITLE
Syncoid does not create identical clones?

### ISSUE THREAD

**Issue by segdy** [jimsalterjrs/sanoid#620]
I have a pool which is currently on a 3x4TB raidz and was originally created many years ago on FreeNAS/FreeBSD:

```
# zfs list
NAME                          USED  AVAIL     REFER  MOUNTPOINT
plvl5i0                      4.40T  2.61T      382M  /plvl5i0
plvl5i0/dsmusic0             1.04T  2.61T     1.04T  /plvl5i0/dsmusic0
plvl5i0/radix                 575K  2.61T      192K  /plvl5i0/radix
plvl5i0/radix/data            298K  2.61T      213K  /zpradix1imain/data
plvl5i0/zvdata0ecsd          1.79T  3.37T      877G  -
plvl5i0/zvhome0ecsd           778G  2.86T      258G  -
plvl5i0/zvsys1ecsd            833G  2.71T      140G  -
#
```

This pool I would like to clone onto a new one consisting of 1x12TB (for now). Like this:

```
# zpool create -o ashift=12 newpool /dev/disk/by-id/ata-TOSHIBA_...
# syncoid -r plvl5i0 newpool/plvl5i0
```

This command runs successfully but the two pools are not identical:

```
# zfs list
NAME                          USED  AVAIL     REFER  MOUNTPOINT
newpool                      3.35T  7.21T       96K  /newpool
newpool/plvl5i0              3.35T  7.21T      400M  /newpool/plvl5i0
newpool/plvl5i0/dsmusic0     1.07T  7.21T     1.07T  /newpool/plvl5i0/dsmusic0
newpool/plvl5i0/radix         360K  7.21T       88K  /newpool/plvl5i0/radix
newpool/plvl5i0/radix/data    200K  7.21T       88K  /newpool/plvl5i0/radix/data
newpool/plvl5i0/zvdata0ecsd  1.04T  7.21T      878G  -
newpool/plvl5i0/zvhome0ecsd   530G  7.21T      258G  -
newpool/plvl5i0/zvsys1ecsd    738G  7.21T      140G  -
plvl5i0                      4.40T  2.61T      382M  /plvl5i0
plvl5i0/dsmusic0             1.04T  2.61T     1.04T  /plvl5i0/dsmusic0
plvl5i0/radix                 575K  2.61T      192K  /plvl5i0/radix
plvl5i0/radix/data            298K  2.61T      213K  /zpradix1imain/data
plvl5i0/zvdata0ecsd          1.79T  3.37T      877G  -
plvl5i0/zvhome0ecsd           778G  2.86T      258G  -
plvl5i0/zvsys1ecsd            833G  2.71T      140G  -
```

As can be seen, all of the USED, AVAIL, REFER columns are different.
And overall, the new pool is over 1TB smaller.

I am worried that syncoid does not do actual clones and forgets some data. I compared all datasets (via `zfs list -t all -r`) and the number of datasets are the same. Yet every single number is different.

Can I trust that **all** data (incl. snapshots, zvols, properties, ...) has been cloned successfully?
If yes, how come that there is all this discrepancy?

PS: I am not sure if this actually a bug but it feels pretty confusing to me and maybe something that can be mentioned in the documentation.

**Comment by phreaker0** [comment]
@segdy this is not a syncoid issue. This is just how ZFS works. The space consumption depends on multiple things: vdev layout, compression properties, ashift, ... .
If you compare the snapshot GUID you should be fine.

For example your src uses a raidz vdev which has more space overhead than a single disk/mirror vdev as used as your target.

**Comment by jimsalterjrs** [comment]
Demonstration:

~~~~
root@banshee:/tmp# truncate -s 1G oldpool0.bin
root@banshee:/tmp# rm oldpool0.bin
root@banshee:/tmp# truncate -s 1G oldpool0.bin
root@banshee:/tmp# truncate -s 1G oldpool1.bin
root@banshee:/tmp# truncate -s 1G oldpool2.bin
root@banshee:/tmp# truncate -s 1G newpool.bin
root@banshee:/tmp# zpool create -oashift=12 oldpool raidz1 /tmp/oldpool0.bin /tmp/oldpool1.bin /tmp/oldpool2.bin
root@banshee:/tmp# zpool create -oashift=12 newpool /tmp/newpool.bin
~~~~

Now we've got an `oldpool` with 3-drive RAIDz1, and a `newpool` with a single-disk vdev. Let's copy data identically to both:

~~~~
root@banshee:/tmp# cp -a /lib/modules/5.4.0-58-generic /oldpool/
root@banshee:/tmp# cp -a /lib/modules/5.4.0-58-generic /newpool/
~~~~

zfs list will show different sizes for the two pools with identical data, because zfs list measures on-disk usage after parity/redundancy, not the number of bytes of actual data:

~~~~
root@banshee:/tmp# zfs list oldpool ; zfs list newpool
NAME      USED  AVAIL     REFER  MOUNTPOINT
oldpool   298M  1.42G      298M  /oldpool
NAME      USED  AVAIL     REFER  MOUNTPOINT
newpool   285M   546M      285M  /newpool
~~~~

But `du -hs --apparent-size` will show the same values for both:

~~~~
root@banshee:/tmp# du -hs --apparent-size /oldpool /newpool
247M    /oldpool
247M    /newpool
~~~~

And so does `md5sum`:

~~~~
root@banshee:/tmp# find /lib/modules/5.4.0-58-generic -type f | sort | xargs -I% cat % | md5sum
2cc136c58ca2b5cc620edac530996963  -
root@banshee:/tmp# find /oldpool/5.4.0-58-generic -type f | sort | xargs -I% cat % | md5sum
2cc136c58ca2b5cc620edac530996963  -
root@banshee:/tmp# find /newpool/5.4.0-58-generic -type f | sort | xargs -I% cat % | md5sum
2cc136c58ca2b5cc620edac530996963  -
~~~~

TL;DR `zfs list` reported USED will be different on different topologies, even for the same underlying data.

**Comment by segdy** [comment]
That makes totally sense now.
Thank you for elaborating
